applications/ConvectionDiffusionApplication/custom_utilities/embedded_mls_constraint_process.cpp

```diff
@@ -0,0 +1,437 @@
+        auto& r_geom = rElement.GetGeometry();
+        if (IsSplit(r_geom)) {
+            // Mark the intersected elements as BOUNDARY
+            // The intersected elements are also flagged as non ACTIVE to avoid assembling them
+            // Note that the split elements BC is applied by means of the extension operators
+            rElement.Set(ACTIVE, false);
+            rElement.Set(BOUNDARY, true);
+            // Mark the positive intersected nodes as BOUNDARY
+            for (auto& rNode : r_geom) {
+                if (!(rNode.FastGetSolutionStepValue(DISTANCE) < 0.0)) {
+                    rNode.Set(BOUNDARY, true);
+                }
+            }
+        } else if (!IsNegative(r_geom)) {
+            // Mark the positive element as ACTIVE to assemble it
+            rElement.Set(ACTIVE, true);
+            // Mark the active element nodes as ACTIVE
+            for (auto& rNode : r_geom) {
+                rNode.Set(ACTIVE, true);
+            }
+        }
+    }
+
+    // Find the surrogate boundary elements
+    // Note that we rely on the fact that the neighbours are sorted according to the faces
+    for (auto& rElement : mpModelPart->Elements()) {
+        if (rElement.Is(BOUNDARY)) {
+            const auto& r_geom = rElement.GetGeometry();
+            const std::size_t n_faces = r_geom.FacesNumber();
+            auto& r_neigh_elems = rElement.GetValue(NEIGHBOUR_ELEMENTS);
+            for (std::size_t i_face = 0; i_face < n_faces; ++i_face) {
+                // The neighbour corresponding to the current face is ACTIVE means that the current face is surrogate boundary
+                // Flag the current neighbour owning the surrogate face as INTERFACE
+                auto& r_neigh_elem = r_neigh_elems[i_face];
+                if (r_neigh_elem.Is(ACTIVE)) {
+                    r_neigh_elem.Set(INTERFACE, true);
+                }
+            }
+        }
+    }
+}
+
+void EmbeddedMLSConstraintProcess::ReactivateElementsAndNodes()
+{
+    // Make intersected elements and their nodes ACTIVE
+    if ( !mDeactivateIntersectedElements ) {
+        for (auto& rElement : mpModelPart->Elements()) {
+            // Check if the element is split
+            const auto& r_geom = rElement.GetGeometry();
+            if (IsSplit(r_geom)) {
+                rElement.Set(ACTIVE, true);
+                for (auto& rNode : r_geom) {
+                    rNode.Set(ACTIVE, true);
+                }
+            }
+        }
+    }
+    // Make negative elements and their nodes ACTIVE
+    if ( !mDeactivateNegativeElements ) {
+        for (auto& rElement : mpModelPart->Elements()) {
+            // Check if the element is negative
+            const auto& r_geom = rElement.GetGeometry();
+            if (IsNegative(r_geom)) {
+                rElement.Set(ACTIVE, true);
+                for (auto& rNode : r_geom) {
+                    rNode.Set(ACTIVE, true);
+                }
+            }
+        }
+    }

```

Comment:
Should I make those for loops parallel as well? With a critical section inside the if clauses?